### PR TITLE
feat: band mask created by hand 

### PR DESCRIPTION
**Pull request by Fersoil**
Simplified the flow of creating a band mask - for now the core feature of our package, since spectral analysis is something new.

To create a band mask, users may pass list of bands, ranges of wavelengths, ranges of indices of wavelengths and dictionary containing simple list of indices. 

Band names are matched using `spyndex` library

Created also more detailed tests

I guess, I could describe this feature more thoroughly - create a diagram, some use-cases or better description

### CODE REVIEW

File: src/meteors/lime.py [hunk 1]
```diff
@@ -264,139 +267,407 @@ def get_segmentation_mask(
...
-            if isinstance(segment, str):
-                segment = [segment]
-            segment = tuple(segment)
-
-            band_names_dict[segment] = idx + 1
-        return band_names_dict
+    def __make_band_names_hashable(
+        segment_name: Sequence[str] | str,
+    ) -> tuple[str, ...] | str:
+        if isinstance(segment_name, str):
```
[the remark this answers is not in the document]
Comment (WolodjaZ) [reply src/meteors/lime.py:342]: Okay I think we can do type validation in private functions only If not:
`
- the code is performance-sensitive, additional type checks could introduce unnecessary overhead.
- the public methods thoroughly check types and there is no reasonable way incorrect types could be passed to private methods, additional checks might be redundant.
`

File: src/meteors/lime.py [hunk 2]
```diff
@@ -16,6 +16,8 @@
 import numpy as np
 import spyndex
 
+import loguru
```
[the remark this answers is not in the document]
Comment (Fersoil) [reply src/meteors/lime.py:19]: As far as I remember `kw_only` in `Field` was added by you, I don't really know if we need to use this parameter

[the remark this answers is not in the document]
Comment (Fersoil) [reply src/meteors/lime.py:19]: `loguru.logger` changed for `logger`

File: src/meteors/lime.py [hunk 3]
```diff
@@ -264,139 +267,407 @@ def get_segmentation_mask(
     @staticmethod
     def get_band_mask(
         image: Image,
-        band_names: list[str | Sequence[str]] | dict[tuple[str, ...] | str, int],
```
[the remark this answers is not in the document]
Comment (Fersoil) [reply src/meteors/lime.py:267]: changed each `Sequence` for `list`

File: src/meteors/lime.py [hunk 4]
```diff
@@ -264,139 +267,407 @@ def get_segmentation_mask(
...
-        }
-
-        return (
-            Lime.__get_band_mask_from_names_dict(image, band_names_dict),  # type: ignore
-            band_names_simplified,  # type: ignore
+        if band_ranges_indices is not None:
+            loguru.logger.debug(
+                "Getting band mask from band groups given by ranges of indices"
+            )
+            band_groups = Lime.__get_band_groups_from_band_ranges_indices(
```
[the remark this answers is not in the document]
Comment (Fersoil) [reply src/meteors/lime.py:322]: Modified to more open approach with single underscore

File: src/meteors/lime.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Fersoil) [reply src/meteors/lime.py:342]: I believe that public functions should do the checks, the performance aspect is also relevant

File: src/meteors/lime.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (WolodjaZ) [reply src/meteors/lime.py]: Interesting I have never used `kw_only` so lets remove it.

**Review by WolodjaZ — APPROVED** [review]

File: src/meteors/lime.py [hunk 5]
```diff
@@ -241,156 +252,453 @@ def to(self, device: torch.device) -> Self:
...
+            if isinstance(band_names, dict):
+                if not all(isinstance(key, (tuple, str)) for key in band_names.keys()):
+                    raise TypeError("Keys in band names dictionary should be either a tuple or a string")
+                if not all(isinstance(value, int) for value in band_names.values()):
+                    raise TypeError("Values in band names dictionary should be integers")
+            if isinstance(band_names, list):
+                if not all(isinstance(segment, (str, list)) for segment in band_names):
+                    raise TypeError("Elements in band names list should be either a string or a list")
+
+            band_ranges_wavelengths, dict_labels_to_segment_ids = Lime._get_band_ranges_wavelengths_from_band_names(
```
Comment (WolodjaZ) [inline src/meteors/lime.py:316]: Mypy gives me:
`src/meteors/lime.py:316: error: Incompatible types in assignment (expression has type "dict[Union[str, tuple[str, ...]], list[tuple[float, float]]]", variable has type "Optional[dict[Union[str, tuple[str, ...]], Union[Iterable[tuple[float, float]], tuple[float, float]]]]")  [assignment]
Found 1 error in 1 file (checked 12 source files)`

File: src/meteors/lime.py [hunk 6]
```diff
@@ -241,156 +252,453 @@ def to(self, device: torch.device) -> Self:
...
+                    f"Invalid band name {band_name}, band name must be either in `spyndex.indices` or `spyndex.bands`"
+                )
+
+        return tuple(band_names_segment)
 
-        device = image.image.device
-        resolution_segments = Lime.__get_resolution_segments(image.wavelengths, grouped_band_names, device=device)
+    @staticmethod
+    def _get_band_ranges_wavelengths_from_band_names(
+        image: Image,
```
Comment (WolodjaZ) [inline src/meteors/lime.py:398]: variable image is not used in the function

**Review by WolodjaZ — CHANGES_REQUESTED** [review]

File: src/meteors/lime.py (hunk 5 above)
Comment (WolodjaZ) [reply src/meteors/lime.py:316]: ignoring as @Fersoil decided

File: src/meteors/lime.py (hunk 6 above)
Comment (WolodjaZ) [reply src/meteors/lime.py:398]: @Fersoil still the variable image is not used.

File: src/meteors/lime.py [hunk 7]
```diff
@@ -241,156 +252,453 @@ def to(self, device: torch.device) -> Self:
...
-        band_mask = resolution_segments.unsqueeze(axis[0]).unsqueeze(axis[1])
-        size_image = image.image.size()
-        size_mask = band_mask.size()
+        Returns:
+            dict[str | tuple[str, ...], list[tuple[float, float]]], dict[str | tuple[str, ...], int]: tuple containing the dictionary with mapping segment labels into wavelength ranges and mapping from segment labels into segment ids
+        """
 
-        repeat_dims = [s2 // s1 for s1, s2 in zip(size_mask, size_image)]
-        band_mask = band_mask.repeat(repeat_dims)
+        segments_list = []
```
Comment (WolodjaZ) [inline src/meteors/lime.py:414]: @Fersoil is there a need for empty list initialization if it is always assigned?

**Review by WolodjaZ — CHANGES_REQUESTED** [review]

**Review by WolodjaZ — APPROVED** [review]

**Pull request merged**